src/Scheduler.scala

```diff
@@ -0,0 +1,40 @@
+import java.util.concurrent.{Executors, PriorityBlockingQueue}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+//type Task = () => Unit
+
+trait Scheduler {
+  def schedule(task: () => Unit, timestamp: Long): Unit
+  def start()(implicit ec: ExecutionContext): Unit
+}
+
+class SchedulerImp(
+    val minHeap: PriorityBlockingQueue[ScheduledTask],
+    // ExecutionContext for Future to execute a task in worker thread (not calling thread)
+    implicit val ec: ExecutionContext
+) extends Scheduler {
+
+  override def schedule(task: () => Unit, timestamp: Long): Unit = {
+    minHeap.offer(ScheduledTask(task, timestamp))
+  }
+
+  override def start()(implicit ec: ExecutionContext): Unit = {
+    Future {
+      while (true) {
+        // put and take are blocking
+        val nextTask = minHeap.take()
+        if (nextTask.timestamp < System.currentTimeMillis) {
+          Future {
+            nextTask.task()
+          }
+        } else {
+          minHeap.put(nextTask)
+          Thread.sleep(100)
```

Comment:
introducing short sleep could alleviate this busy spinning. but, eventually we need to find better way.